### PR TITLE
doc/rgw/account: Handling notification topics when migrating an existing user into an account

### PR DESCRIPTION
**Pull request by BBoozmen**
Add a subsection under "Migrate an existing User into an Account" to describe how a client can seamlessly migrate the notification topics after account migration.

Fixes https://tracker.ceph.com/issues/67656

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by BBoozmen** [comment]
jenkins test docs

File: doc/radosgw/account.rst [hunk 1]
```diff
@@ -252,3 +287,5 @@ This example uses `awscli`_ to create an IAM user for S3 operations.
 .. _Evaluating policies within a single account: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic.html#policy-eval-basics
 .. _Cross-account policy evaluation logic: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_evaluation-logic-cross-account.html
 .. _awscli: https://docs.aws.amazon.com/cli/latest/
+.. _Bucket Notifications: ../notifications/
+.. _Supported Zone Features: ../zone-features/supported-features
```
Comment (cbodley) [inline doc/radosgw/account.rst:291]: this one is a broken link (https://ceph--59390.org.readthedocs.build/en/59390/radosgw/zone-features/supported-features returns 404)

i wonder if `../zone-features/#supported-features` would work? if not i think it would be fine to link to `../zone-features` since "supported features" is near the top

File: doc/radosgw/account.rst [hunk 2]
```diff
@@ -187,6 +187,41 @@ Alternatively, you may want to create a new account for each existing user. In
...
+
+    .. code-block:: json
+
+        {"TopicConfigurations": [{ "Id": "ID1", "TopicArn": "arn:aws:sns:default:RGW00000000000000001:topic1", "Events": ["s3:ObjectCreated:*"]}]}
+
+In this example, `RGW00000000000000001` is the account ID, `topic1` is the
+topic name and `ID1` is the notification ID.
+
+4. ``Removing Old Topics``: Once no buckets are subscribed to the old user-owned topics,
+they can be removed using the "topic rm" command via radosgw-admin.
```
Comment (cbodley) [inline doc/radosgw/account.rst:223]: since we know the topic name in this example, let's spell out the radosgw-admin command?
```rst
4. ``Removing Old Topics``: Once no buckets are subscribed to the old user-owned topics,
they can be removed::
	$ radosgw-admin topic rm --topic topic1
```

**Review by cbodley — COMMENTED** [review]
looks great, very clear instructions 👍

**Comment by adamemerson** [comment]
jenkins retest this please

File: doc/radosgw/account.rst (hunk 1 above)
Comment (BBoozmen) [reply doc/radosgw/account.rst]: `#` did the trick. The link seems to be working now.

File: doc/radosgw/account.rst (hunk 2 above)
Comment (BBoozmen) [reply doc/radosgw/account.rst]: Sounds good. Updated Step 4.

Comment (BBoozmen) [reply doc/radosgw/account.rst]: Yeap, added the second `:` at the end of the sentence. For some reason, my vscode rst viewer treats both versions the same way.

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
any feedback from @ceph/doc-writers? we'd like to get this to squid asap

File: doc/radosgw/account.rst [hunk 3]
```diff
@@ -187,6 +187,43 @@ Alternatively, you may want to create a new account for each existing user. In
 that case, you may want to add the ``--account-root`` option to make each user
 the root user of their account.
 
+Migrating Notification Topics
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Account topics are only supported when ``notification_v2`` feature is enabled,
```
Comment (anthonyeleven) [inline doc/radosgw/account.rst:193]: are supported only when the ``notification_v2`` feature

File: doc/radosgw/account.rst [hunk 4]
```diff
@@ -187,6 +187,43 @@ Alternatively, you may want to create a new account for each existing user. In
...
+as described in `Bucket Notifications`_ and `Supported Zone Features`_.
+
+1. ``Migration Impact``: When a non-account user is migrated to an account,
+the user loses access to its existing notification topics via the SNS Topic API.
+However, ongoing notifications remain unaffected.
```
Comment (anthonyeleven) [inline doc/radosgw/account.rst:198]: I'm unclear what "ongoing" here means vs existing notification topics, to someone not versed in this area this sounds like a contradiction.

File: doc/radosgw/account.rst [hunk 5]
```diff
@@ -187,6 +187,43 @@ Alternatively, you may want to create a new account for each existing user. In
...
+
+2. ``Recreation of Topics``: The account-user should recreate the topics using
```
Comment (anthonyeleven) [inline doc/radosgw/account.rst:200]: I personally write `re-create` when not discussing fun and games, though this is somewhat controversial.

I would take out the hypen between `account` and `user.

**Review by anthonyeleven — APPROVED** [review]
@cbodley I've made a few suggestions, none that warrant blocking, so I've approved but would very much like to see some tweaks.

File: doc/radosgw/account.rst (hunk 4 above)
Comment (cbodley) [reply doc/radosgw/account.rst:198]: maybe "However, bucket notifications on these topics will continue to be delivered."

File: doc/radosgw/account.rst (hunk 3 above)
Comment (BBoozmen) [reply doc/radosgw/account.rst]: Done.

File: doc/radosgw/account.rst (hunk 4 above)
Comment (BBoozmen) [reply doc/radosgw/account.rst]: > maybe "However, bucket notifications on these topics will continue to be delivered."

Just saw this after adding the new evidence but the latest change is on the similar lines:

> However, these topics remain functional, and bucket notifications will continue
to be delivered as expected.

Let me know how it looks.

File: doc/radosgw/account.rst (hunk 5 above)
Comment (BBoozmen) [reply doc/radosgw/account.rst]: Done.

File: doc/radosgw/account.rst (hunk 4 above)
Comment (cbodley) [reply doc/radosgw/account.rst]: even better

File: doc/radosgw/account.rst [hunk 6]
```diff
@@ -187,6 +187,44 @@ Alternatively, you may want to create a new account for each existing user. In
...
+Migrating Notification Topics
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Account topics are supported only when the ``notification_v2`` feature is enabled,
+as described in `Bucket Notifications`_ and `Supported Zone Features`_.
+
+1. ``Migration Impact``: When a non-account user is migrated to an account,
+the user loses access to its existing notification topics via the SNS Topic API.
+However, these topics remain functional, and bucket notifications will continue
+to be delivered as expected.
```
Comment (yuvalif) [inline doc/radosgw/account.rst:199]: maybe add a note that they are still accessible by the admin API?

**Comment by yuvalif** [comment]
@BBoozmen what about topics that were created usign a tenant? does it require a different process?

File: doc/radosgw/account.rst (hunk 6 above)
Comment (BBoozmen) [reply doc/radosgw/account.rst]: Added some more clarification around this.

**Comment by BBoozmen** [comment]
> @BBoozmen what about topics that were created usign a tenant? does it require a different process?

Hmm, I haven't tested this scenario. Let me take a look. However, when naming the rados object (for a topic) since either account or tenant is used, I'm guessing this method may not work??? Let me also check how we migrate a tenanted user under an account.

**Comment by BBoozmen** [comment]
@yuvalif 
> what about topics that were created using a tenant? does it require a different process?

@cbodley 
> the only issue i'm aware of is https://tracker.ceph.com/issues/63018, where tenanted topics didn't namespace the persistent queue object. Other than that preexisting issue, i'd expect the account migration to work fine. the original topic's queue would not have a namespace, but the account topic's queue would have the account-id namespace.

OK, I got a chance to test this scenario. 

TLDR; this procedure is working OK with tenanted users as well. Of course, this is possible when selected [account-ids](https://docs.ceph.com/en/latest/radosgw/account/#account-ids) are unique within the cluster. However, I see one minor issue described below in the very end.

Below is more details:

1. Create a tenanted-user ("tenant1$myuser1") and let it create a topic (using AWS SNS API)
<pre>
./bin/radosgw-admin topic list --tenant=tenant1
{
    "topics": [
        {
            "owner": "tenant1$myuser1",
            "name": "kf-oozmen-user1",
            "dest": {
                ...
                "push_endpoint_topic": "kf-oozmen-user1",
                "persistent_queue": "tenant1:kf-oozmen-user1",
            },
            "arn": "arn:aws:sns:default:tenant1:kf-oozmen-user1",
            ...
            "subscribed_buckets": []
        }
    ]
}
</pre>
2. Create a bucket and put notification cfg to use this topic.
3. Verify notifications are working.
4. Create an account under the same tenant (from step 1) and move the user under this account.
<pre>
$ radosgw-admin account create --tenant=tenant1 --account-name=RGW00000000000000001 --account-id=RGW00000000000000001
$ radosgw-admin user modify --tenant=tenant1 --uid=myuser1 --account-id=RGW00000000000000001 --account-root
</pre>
5. Repeat Step 2: create another topic with the same topic name. This doesn't overwrite the existing rados object for the original topic:
<pre>
$ rados -p default.rgw.meta ls --all  | grep ^topic
topics  topic.tenant1:kf-oozmen-user1
topics  topic.RGW00000000000000001:kf-oozmen-user1
</pre>
and SNS API sees only the new topic. As expected, old topic is not accessible for the user anymore. New topic has its own queue:
<pre>
$ radosgw-admin topic list --account-id=RGW00000000000000001
{
    "topics": [
        {
            "owner": "RGW00000000000000001",
            "name": "kf-oozmen-user1",
            "dest": {
                ...
                "push_endpoint_topic": "kf-oozmen-user1",
                "persistent_queue": "RGW00000000000000001:kf-oozmen-user1",
            },
            "arn": "arn:aws:sns:default:RGW00000000000000001:kf-oozmen-user1",
            "subscribed_buckets": []
        }
    ]
}
</pre>
6. Verify old topic is still functional and no duplicate notifications.
7. Change bucket's notification configuration to use the new topic. Remove tenanted topic and verify notifications are still working.

Everything works just fine. However, the only minor issue is the delivered notification still refers to the old/tenanted ARN. For example, below is the notification received by the client:

<pre>
{"Records":[{...,
    "eventName":"ObjectCreated:Put",
    "userIdentity":{"principalId":"RGW00000000000000001"},....
    "s3":{"configurationId":"mytopic","bucket":{"name":"u1b1","ownerIdentity":{"principalId":"RGW00000000000000001"}, 
      ...
     "arn":"arn:aws:s3:default:tenant1:u1b1", <<<<<<<<<<<<<<<<<<<<<
....}]}
</pre>

While client sees a different ARN (account-scoped as expected) when listing the topic:

<pre>
>>> c1_sns.list_topics()
{'Topics': [{'TopicArn': 'arn:aws:sns:default:RGW00000000000000001:kf-oozmen-user1'}], ...}
</pre>

I believe this is related to how we construct topic-arn at https://github.com/ceph/ceph/blob/15cbae0a2745e79e6a39d694b39e5140428c3148/src/rgw/rgw_rest_pubsub.cc#L302. But, this is not really about this procedure and its applicability for tenanted user; totally a different issue.

In conclusion, I think this procedure applies for tenanted-user case as well.

**Comment by yuvalif** [comment]
@BBoozmen thanks for the detailed analysis! agree, that this is not an issue for the doc.

regarding the issue itself. wil lthat happen only for notifications that were in the queue before we switched to the new topic?
if this is the case, this is a transient issue, that we can probably ignore.

**Comment by BBoozmen** [comment]
> @BBoozmen thanks for the detailed analysis! agree, that this is not an issue for the doc.
> 
> regarding the issue itself. wil lthat happen only for notifications that were in the queue before we switched to the new topic? if this is the case, this is a transient issue, that we can probably ignore.

Looking at the results again: in the notification string, the arn is actually for the BUCKET resource which still refers to "tenant1" (i.e., "arn":"arn:aws:s3:default:tenant1:**u1b1**") but not topic ARN so we should be good.

I'll edit the previous comment to strikethru the statements re "minor issue". So, notification point of view tenanted-user scenario looks good.

**Review by yuvalif — APPROVED** [review]
